### PR TITLE
Query Resolvers

### PR DESCRIPTION
**Pull request by aqeelmozumder**
**3 files have been modified:**

- prisma/index.ts
- resolvers/index.ts
- auth/index.ts

**4 files have been created:**

- resolverUtils.ts
- user.ts
-  course.ts
- schedule.ts

The others files were to keep the branch updated with the main.

Querying me, finduserbyid, courses and schedule will give correct response.

### CODE REVIEW

**Comment by aomi** [comment]
@aqeelmozumder merging main into this branch as the diff seems odd.

**Comment by aqeelmozumder** [comment]
Merged main into this branch

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -76,3 +76,31 @@ jobs:
...
+          # Strip git ref prefix from version
+          VERSION=$(echo "${{ github.ref }}" | sed -e 's,.*/\(.*\),\1,')
+          # Strip "v" prefix from tag name
+          [[ "${{ github.ref }}" == "refs/tags/"* ]] && VERSION=$(echo $VERSION | sed -e 's/^v//')
+          # Use Docker `latest` tag convention
+          [ "$VERSION" == "main" ] && VERSION=latest
+          echo IMAGE_ID=$IMAGE_ID
+          echo VERSION=$VERSION
+          docker tag $IMAGE_NAME $IMAGE_ID:$VERSION
+          docker push $IMAGE_ID:$VERSION
```
Comment (aomi) [inline .github/workflows/ci.yml]: These changes shouldn't be part of your PR so they must have slipped in while you were working?

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -35,4 +34,4 @@ jobs:
           echo IMAGE_ID=$IMAGE_ID
           echo VERSION=$VERSION
           docker tag $IMAGE_NAME $IMAGE_ID:$VERSION
-          docker push $IMAGE_ID:$VERSION
+          docker push $IMAGE_ID:$VERSION
```
Comment (aomi) [inline .github/workflows/docker.yml]: Same thing here with unrelated changes.

File: src/prisma/index.ts [hunk 3]
```diff
@@ -13,3 +13,27 @@ async function lookupUser(username: string) {
   console.log(user);
   return user;
 }
+
+async function lookupId(userid: number) {
```
Comment (aomi) [inline src/prisma/index.ts:17]: Perhaps we either move these to a `src/users/index.ts` or we make these a bit more descriptive?
ie. `findUserById(id: number)`?

File: src/resolvers/index.ts [hunk 4]
```diff
@@ -22,10 +23,27 @@ const noPerms = {
 
 export const resolvers: Resolvers<Context> = {
   Query: {
-    me: (_, _params, ctx) => {
+    me: async (_, _params, ctx) => {
+      let username;
       if (!ctx.session.user) return null;
-      console.log(ctx.session.user.username);
-      return null;
+      else if (!ctx.session.user.username) return null;
```
Comment (aomi) [inline src/resolvers/index.ts:29]: These could be merged using `&&` I think. I didn't know that the `username` property was optional.

File: src/resolvers/index.ts [hunk 5]
```diff
@@ -22,10 +23,27 @@ const noPerms = {
...
+      else username = ctx.session.user.username;
+
+      return await getMe(username);
```
Comment (aomi) [inline src/resolvers/index.ts:32]: You can remove the `let username` and return directly by removing the `else` here and just passing in the `getMe(ctx.session.user.username)`.

File: src/resolvers/resolverUtils.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  preferences: [],
+  active: false,
+};
+
+const failedSchedule = {
+  id: 'Not Found',
+  year: 0,
+  createdAt: Date,
+  courses: [],
+};
```
Comment (aomi) [inline src/resolvers/resolverUtils.ts:19]: These you can have typed I believe.

File: src/resolvers/resolverUtils.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+  const courses = await lookupCourses(term);
+
+  return {
+    CourseID: courses[0].courseSection[0].courseId as unknown as CourseId,
+    hoursPerWeek: courses[0].courseSection[0].hoursPerWeek,
+    capacity: courses[0].courseSection[0].capacity,
+    professors: [],
+    startDate: courses[0].courseSection[0].startDate,
+    endDate: courses[0].courseSection[0].endDate,
+    meetingTimes: [],
```
Comment (aomi) [inline src/resolvers/resolverUtils.ts:61]: I'm curious why it's only returning a single element. Also this could return 0 for which this will throw an error for accessing an undefined element.

File: src/prisma/index.ts [hunk 8]
```diff
@@ -13,3 +13,27 @@ async function lookupUser(username: string) {
...
+}
+
+async function lookupSchedule(year: number | undefined) {
+  const schedule = await prisma.schedule.findUnique({
+    where: {},
+    include: { courseSection: true },
+  });
+
+  return schedule;
+}
```
Comment (aomi) [inline src/prisma/index.ts:39]: Why can year be `undefined` and also not used here?

**Comment by aqeelmozumder** [comment]
@aomi Merge conflicts are Resolved

File: src/resolvers/resolverUtils.ts [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+  };
+}
+
+async function getSchedule(year: number | undefined): Promise<Schedule> {
+  const schedule = await lookupSchedule(year || undefined);
+
+  if (!schedule) return failedSchedule;
+
+  return {
+    id: 'test', // schedule.id.toString not working
```
Comment (aomi) [inline src/resolvers/resolverUtils.ts:71]: `id: `${schedule.id}`

File: src/resolvers/index.ts [hunk 10]
```diff
@@ -22,10 +23,27 @@ const noPerms = {
...
+    },
+    findUserById: async (_, _params, ctx) => {
+      if (!ctx.session.user) return null;
+      else if (!_params.id) return null;
+      else return await getUserByID(+_params.id);
```
Comment (aomi) [inline src/resolvers/index.ts:37]: Don't need the `else` here either. Also you can remove the `_` from params now since the value is being used.

File: src/resolvers/index.ts [hunk 11]
```diff
@@ -22,10 +23,27 @@ const noPerms = {
...
+    },
+    courses: async (_, _params, ctx) => {
+      if (!ctx.session.user) return null;
+      else if (!_params.term) return null;
+      else return [await getCourses(_params.term)];
```
Comment (aomi) [inline src/resolvers/index.ts:42]: This is a bit confusing given `getCourses` is implying a plural (array) of response and there's something odd happening by creating an array here.

File: src/resolvers/index.ts [hunk 12]
```diff
@@ -22,10 +23,27 @@ const noPerms = {
...
+    },
+    schedule: async (_, _params, ctx) => {
+      if (!ctx.session.user) return null;
+      else return getSchedule(_params.year || undefined);
```
Comment (aomi) [inline src/resolvers/index.ts:46]: Probably don't want `|| undefined` used in such a way and make the `getSchedule` handle it or give it something more explict like the current year etc.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (aqeelmozumder) [reply .github/workflows/ci.yml]: I am not sure, these two GitHub workflow files changed after I Merged the main into this branch.

File: src/prisma/index.ts (hunk 3 above)
Comment (aqeelmozumder) [reply src/prisma/index.ts:17]: I think the function names make sense like lookupUser would be looking up a username. Maybe for courses and schedule I could make it findCourses or findSchedule.

File: src/resolvers/resolverUtils.ts (hunk 7 above)
Comment (aqeelmozumder) [reply src/resolvers/resolverUtils.ts:61]: I was trying for a single element. However, I am not sure how to include all the elements. Do I need to run a loop?

File: src/prisma/index.ts (hunk 3 above)
Comment (LacroixJ) [reply src/prisma/index.ts:17]: I prefer findUserByID. You can also refactor lookupUser to findUserByUsername(and move it too).

You could consider making another file within the prisma folder for utility functions `src/prisma/utils.ts` or even split up all your functions into various subfiles for  similar functions like `prisma/users.ts`, `prisma/schedules.ts` and so on.

File: src/resolvers/resolverUtils.ts [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+};
+
+const failedSchedule = {
+  id: 'Not Found',
+  year: 0,
+  createdAt: Date,
+  courses: [],
+};
+
+async function getMe(username: string): Promise<User> {
```
Comment (LacroixJ) [inline src/resolvers/resolverUtils.ts:21]: You should pass in the user object from the session and parse out from that instead of doing another database lookup.

Pass in the ctx to this instead of the username and get the user with `ctx.session.user`

**Review by LacroixJ — CHANGES_REQUESTED** [review]
Added a couple comments, Aomi hit the main points.

File: src/resolvers/index.ts (hunk 4 above)
Comment (aqeelmozumder) [reply src/resolvers/index.ts:29]: && doesn't work as one is dependant on the other

Comment (aomi) [reply src/resolvers/index.ts:29]: Is the type checker telling you this? It's pretty smart about checking this iirc?

Comment (aqeelmozumder) [reply src/resolvers/index.ts:28]: Yes I think, when i do if (!ctx.session.user && **!ctx.session.user**.username ) return null; The bolded part is the error as it says "Object is possibly 'undefined'"

Comment (aomi) [reply src/resolvers/index.ts:28]: I tried it and ran `npx tsc --noEmit` which runs TypeScript type-checker without emitting and build artifacts and it was fine.
```ts
    me: async (_, _params, ctx) => {
      if (!ctx.session.user || !ctx.session.user.username) return null;
      return await getMe(ctx);
    },
```

Comment (aomi) [reply src/resolvers/index.ts:28]: That being said, I don't think session.user.username` can ever be undefined if `user` is defined.

File: src/prisma/course.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
+import { PrismaClient, Term } from '@prisma/client';
+export { findCourses, findMeetingTime };
+
+const prisma = new PrismaClient();
```
Comment (LacroixJ) [inline src/prisma/course.ts:4]: There must be a better way of doing this, we should not have 3 prisma clients.

The index.ts file should be the module entrypoint

Comment (aqeelmozumder) [reply src/prisma/course.ts:4]: Is it better to export the prisma from seed.ts?
Then we can use one client.

Comment (LacroixJ) [reply src/prisma/course.ts:4]: Seed.ts does not run when the application is running. 

Just put prisma in the `index.ts` and use it via the other files in the module? Or is that not a thing?

Comment (aqeelmozumder) [reply src/prisma/course.ts:4]: I changed the index.ts to user.ts.

Will create a new index.ts to include
export const prisma =new PrismaClient() 

Then other files will use this prisma from index.ts

Comment (LacroixJ) [reply src/prisma/course.ts:4]: That should work.

File: src/resolvers/resolverUtils.ts (hunk 6 above)
Comment (aqeelmozumder) [reply src/resolvers/resolverUtils.ts:34]: Typed did not work

File: .github/workflows/ci.yml (hunk 1 above)
Comment (aqeelmozumder) [reply .github/workflows/ci.yml]: Let me know if i should resolve this conversation or is there anything I have to fix?

**Review by aomi — APPROVED** [review]

**Review by LacroixJ — APPROVED** [review]
